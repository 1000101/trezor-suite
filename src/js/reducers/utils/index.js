/* @flow */


import * as LogActions from '~/js/actions/LogActions';
import * as STORAGE from '~/js/actions/constants/localStorage';
import * as WALLET from '~/js/actions/constants/wallet';
import BigNumber from 'bignumber.js';

import type {
    Middleware,
    MiddlewareAPI,
    MiddlewareDispatch,
    State,
    Dispatch,
    Action,
    AsyncAction,
    GetState,
    Device,
    TrezorDevice,
    Account,
    Coin,
    Discovery,
    Token,
    PendingTx,
    Web3Instance,
} from '~/flowtype';

export const getSelectedDevice = (state: State): ?TrezorDevice => {
    const locationState = state.router.location.state;
    if (!locationState.device) return undefined;

    const instance: ?number = locationState.deviceInstance ? parseInt(locationState.deviceInstance) : undefined;
    return state.devices.find((d) => {
        if (!d.features && d.path === locationState.device) {
            return true;
        } if (d.features && d.features.bootloader_mode && d.path === locationState.device) {
            return true;
        } if (d.features && d.features.device_id === locationState.device && d.instance === instance) {
            return true;
        }
        return false;
    });
};

//
export const isSelectedDevice = (current: ?TrezorDevice, device: ?TrezorDevice): boolean => !!((current && device && (current.path === device.path && current.instance === device.instance)));

// find device by id and state
export const findDevice = (devices: Array<TrezorDevice>, deviceId: string, deviceState: string, instance: ?number): ?TrezorDevice => devices.find((d) => {
    // TODO: && (instance && d.instance === instance)
    if (d.features && d.features.device_id === deviceId && d.state === deviceState) {
        return true;
    }
    return false;
});

// get next instance number
export const getDuplicateInstanceNumber = (devices: Array<TrezorDevice>, device: Device | TrezorDevice): number => {
    // find device(s) with the same features.device_id
    // and sort them by instance number
    const affectedDevices: Array<TrezorDevice> = devices.filter(d => d.features && device.features && d.features.device_id === device.features.device_id)
        .sort((a, b) => {
            if (!a.instance) {
                return -1;
            }
            return !b.instance || a.instance > b.instance ? 1 : -1;
        });
    
    // calculate new instance number
    const instance: number = affectedDevices.reduce((inst, dev) => (dev.instance ? dev.instance + 1 : inst + 1), 0);
    return instance;
};

export const getSelectedAccount = (state: State): ?Account => {
    const device = state.wallet.selectedDevice;
    const locationState = state.router.location.state;
    if (!device || !locationState.network || !locationState.account) return null;

    const index: number = parseInt(locationState.account);

    return state.accounts.find(a => a.deviceState === device.state && a.index === index && a.network === locationState.network);
};

export const getSelectedNetwork = (state: State): ?Coin => {
    const device = state.wallet.selectedDevice;
    const coins = state.localStorage.config.coins;
    const locationState = state.router.location.state;
    if (!device || !locationState.network) return null;

    return coins.find(c => c.network === locationState.network);
};

export const getDiscoveryProcess = (state: State): ?Discovery => {
    const device = state.wallet.selectedDevice;
    const locationState = state.router.location.state;
    if (!device || !locationState.network) return null;

    return state.discovery.find(d => d.deviceState === device.state && d.network === locationState.network);
};

export const getAccountPendingTx = (pending: Array<PendingTx>, account: ?Account): Array<PendingTx> => {
    const a = account;
    if (!a) return [];
    return pending.filter(p => p.network === a.network && p.address === a.address);
};

export const getPendingNonce = (pending: Array<PendingTx>): number => pending.reduce((value: number, tx: PendingTx) => {
    if (tx.rejected) return value;
    return Math.max(value, tx.nonce + 1);
}, 0);

export const getPendingAmount = (pending: Array<PendingTx>, currency: string, token: boolean = false): BigNumber => pending.reduce((value: BigNumber, tx: PendingTx) => {
    if (tx.currency === currency && !tx.rejected) {
        return new BigNumber(value).plus(token ? tx.amount : tx.total);
    }
    return value;
}, new BigNumber('0'));

export const getAccountTokens = (state: State, account: ?Account): Array<Token> => {
    const a = account;
    if (!a) return [];
    return state.tokens.filter(t => t.ethAddress === a.address && t.network === a.network && t.deviceState === a.deviceState);
};

export const getWeb3 = (state: State): ?Web3Instance => {
    const locationState = state.router.location.state;
    if (!locationState.network) return null;
    return state.web3.find(w3 => w3.network === locationState.network);
};
