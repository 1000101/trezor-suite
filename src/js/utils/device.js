import colors from 'js/config/colors';

const getStatus = (device) => {
    let status = 'connected';
    if (!device.connected) {
        status = 'disconnected';
    } else if (!device.available) {
        status = 'unavailable';
    } else if (device.type === 'acquired') {
        if (device.status === 'occupied') {
            status = 'used-in-other-window';
        }
    } else if (device.type === 'unacquired') {
        status = 'unacquired';
    }

    return status;
};

const getStatusName = (deviceStatus) => {
    let statusName;
    switch (deviceStatus) {
        case 'used-in-other-window':
            statusName = 'Used in other window';
            break;
        case 'connected':
            statusName = 'Connected';
            break;
        case 'disconnected':
            statusName = 'Disconnected';
            break;
        case 'unacquired':
            statusName = 'Used in other window';
            break;
        case 'unavailable':
            statusName = 'Unavailable';
            break;
        default:
            statusName = 'Status unknown';
    }
    return statusName;
};

const isWebUSB = transport => !!((transport && transport.version.indexOf('webusb') >= 0));

const isDisabled = (selectedDevice, devices, transport) => (devices.length < 1 && !isWebUSB(transport)) || (devices.length === 1 && !selectedDevice.features && !isWebUSB(transport));

const getVersion = (device) => {
    let version;
    if (device.features && device.features.major_version > 1) {
        version = 'T';
    } else {
        version = '1';
    }
    return version;
};

const getStatusColor = (deviceStatus) => {
    let color;
    switch (deviceStatus) {
        case 'used-in-other-window':
            color = colors.WARNING_PRIMARY;
            break;
        case 'connected':
            color = colors.GREEN_PRIMARY;
            break;
        case 'unacquired':
            color = colors.WARNING_PRIMARY;
            break;
        case 'disconnected':
            color = colors.ERROR_PRIMARY;
            break;
        case 'unavailable':
            color = colors.ERROR_PRIMARY;
            break;
        default:
            color = colors.TEXT_PRIMARY;
    }
    return color;
};

export {
    getStatus,
    isDisabled,
    getStatusName,
    getVersion,
    getStatusColor,
};