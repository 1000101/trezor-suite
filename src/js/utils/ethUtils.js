/* @flow */
'use strict';

import BigNumber from 'bignumber.js'

export const decimalToHex = (dec: number): string => {
    return new BigNumber(dec).toString(16);
}

export const hexToDecimal = (hex: number): string => {
    return new BigNumber(this.sanitizeHex(hex)).toString();
}

export const sanitizeHex = (hex: ?string): string => {
    if (typeof hex !== 'string') return null;
    hex = hex.substring(0, 2) === '0x' ? hex.substring(2) : hex;
    if (hex === '') return '';
    return '0x' + padLeftEven(hex);
}

export const padLeftEven = (hex) => {
    hex = hex.length % 2 != 0 ? '0' + hex : hex;
    return hex;
}

export const strip = (str: string): string => {
    if (str.indexOf('0x') === 0) {
        return padLeftEven( str.substring(2, str.length) );
    }
    return padLeftEven(str);
}

export const calcGasPrice = (price: BigNumber, limit): string => {
    return price.times(limit);
}