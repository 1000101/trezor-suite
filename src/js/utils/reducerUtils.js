/* @flow */
'use strict';


export const getAccounts = (accounts: Array<any>, device: any, network: ?string): Array<any> => {
    if (network) {
        return accounts.filter((addr) => addr.deviceState === device.state && addr.network === network);
    } else {
        return accounts.filter((addr) => addr.deviceState === device.state);
    }
    
}

// Public method used in components to find device by state and device_id
export const findDevice = (devices: Array<TrezorDevice>, state: ?string, deviceId: ?string, instance: ?string): ?TrezorDevice => {
    return devices.find(d => d.state === state && d.features && d.features.device_id === deviceId && d.instance === instance);
}