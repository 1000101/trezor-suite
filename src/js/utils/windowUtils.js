/* @flow */
'use strict';

export const getViewportHeight = (): number => (
    window.innerHeight
      || document.documentElement.clientHeight
      || document.body.clientHeight
)

export const getScrollY = (): number => {
    if (window.pageYOffset !== undefined) {
        return window.pageYOffset;
    } else if (window.scrollTop !== undefined) {
        return window.scrollTop;
    } else {
        return (document.documentElement || document.body.parentNode || document.body).scrollTop;
    }
}