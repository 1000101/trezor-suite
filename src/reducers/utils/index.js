/* @flow */
import BigNumber from 'bignumber.js';

import type {
    State,
    Device,
    TrezorDevice,
    Account,
    Coin,
    Discovery,
    Token,
    PendingTx,
    Web3Instance,
} from 'flowtype';

export const getSelectedDevice = (state: State): ?TrezorDevice => {
    const locationState = state.router.location.state;
    if (!locationState.device) return undefined;

    const instance: ?number = locationState.deviceInstance ? parseInt(locationState.deviceInstance, 10) : undefined;
    return state.devices.find((d) => {
        if (!d.features && d.path === locationState.device) {
            return true;
        } if (d.features && d.features.bootloader_mode && d.path === locationState.device) {
            return true;
        } if (d.features && d.features.device_id === locationState.device && d.instance === instance) {
            return true;
        }
        return false;
    });
};

//
export const isSelectedDevice = (current: ?TrezorDevice, device: ?TrezorDevice): boolean => !!((current && device && (current.path === device.path && current.instance === device.instance)));

// find device by id and state
export const findDevice = (devices: Array<TrezorDevice>, deviceId: string, deviceState: string, instance: ?number): ?TrezorDevice => devices.find((d) => {
    // TODO: && (instance && d.instance === instance)
    if (d.features && d.features.device_id === deviceId && d.state === deviceState) {
        return true;
    }
    return false;
});

// get next instance number
export const getDuplicateInstanceNumber = (devices: Array<TrezorDevice>, device: Device | TrezorDevice): number => {
    // find device(s) with the same features.device_id
    // and sort them by instance number
    const affectedDevices: Array<TrezorDevice> = devices.filter(d => d.features && device.features && d.features.device_id === device.features.device_id)
        .sort((a, b) => {
            if (!a.instance) {
                return -1;
            }
            return !b.instance || a.instance > b.instance ? 1 : -1;
        });

    // calculate new instance number
    const instance: number = affectedDevices.reduce((inst, dev) => (dev.instance ? dev.instance + 1 : inst + 1), 0);
    return instance;
};

export const getSelectedAccount = (state: State): ?Account => {
    const device = state.wallet.selectedDevice;
    const locationState = state.router.location.state;
    if (!device || !locationState.network || !locationState.account) return null;

    const index: number = parseInt(locationState.account, 10);

    return state.accounts.find(a => a.deviceState === device.state && a.index === index && a.network === locationState.network);
};

export const getSelectedNetwork = (state: State): ?Coin => {
    const device = state.wallet.selectedDevice;
    const { coins } = state.localStorage.config;
    const locationState = state.router.location.state;
    if (!device || !locationState.network) return null;

    return coins.find(c => c.network === locationState.network);
};

export const getDiscoveryProcess = (state: State): ?Discovery => {
    const device = state.wallet.selectedDevice;
    const locationState = state.router.location.state;
    if (!device || !locationState.network) return null;

    return state.discovery.find(d => d.deviceState === device.state && d.network === locationState.network);
};

export const getAccountPendingTx = (pending: Array<PendingTx>, account: ?Account): Array<PendingTx> => {
    const a = account;
    if (!a) return [];
    return pending.filter(p => p.network === a.network && p.address === a.address);
};

export const getPendingNonce = (pending: Array<PendingTx>): number => pending.reduce((value: number, tx: PendingTx) => {
    if (tx.rejected) return value;
    return Math.max(value, tx.nonce + 1);
}, 0);

export const getPendingAmount = (pending: Array<PendingTx>, currency: string, token: boolean = false): BigNumber => pending.reduce((value: BigNumber, tx: PendingTx) => {
    if (tx.currency === currency && !tx.rejected) {
        return new BigNumber(value).plus(token ? tx.amount : tx.total);
    }
    return value;
}, new BigNumber('0'));

export const getAccountTokens = (state: State, account: ?Account): Array<Token> => {
    const a = account;
    if (!a) return [];
    return state.tokens.filter(t => t.ethAddress === a.address && t.network === a.network && t.deviceState === a.deviceState);
};

export const getWeb3 = (state: State): ?Web3Instance => {
    const locationState = state.router.location.state;
    if (!locationState.network) return null;
    return state.web3.find(w3 => w3.network === locationState.network);
};

export const observeChanges = (prev: ?(Object | Array<any>), current: ?(Object | Array<any>), fields?: Array<string>): boolean => {
    if (prev !== current) {
        // 1. one of the objects is null/undefined
        if (!prev || !current) return true;
        // 2. object are Arrays and they have different length
        if (Array.isArray(prev) && Array.isArray(current)) return prev.length !== current.length;
        // 3. no nested field to check
        if (!Array.isArray(fields)) return true;
        // 4. validate nested field
        if (prev instanceof Object && current instanceof Object) {
            for (let i = 0; i < fields.length; i++) {
                const key = fields[i];
                if (prev[key] !== current[key]) return true;
            }
        }
    }
    return false;
};
