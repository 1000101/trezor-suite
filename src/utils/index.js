import Version from './Version';
import FirmwareInfo from './FirmwareInfo';
import T1 from './T1';

export {
    Version,
    FirmwareInfo,
    T1,
};
