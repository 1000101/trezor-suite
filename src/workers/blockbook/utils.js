const transformAccountInfo = (payload, details) => {
    const result = payload;

    if (details === 'basic') {
        result.transactions = [];
        result.tokens = [];
    }

    return result.sort();
};

export { transformAccountInfo };
