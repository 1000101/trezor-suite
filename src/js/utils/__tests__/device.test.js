import * as dUtils from 'utils/device';

describe('device utils', () => {
    it('get status', () => {
        const deviceMock = [
            {
                device: {
                    connected: false,
                    type: 'acquired',
                    status: 'occupied',
                },
            },
            { device: { connected: true } },
            { device: { available: false } },
            { device: { available: true } },
        ];

        deviceMock.forEach((device) => {
            expect(dUtils.getStatusName(device)).toMatchSnapshot();
        });
    });

    it('get status color', () => {
        const entry = [
            0,
            null,
            'sdsdsdsd',
            'used-in-other-window',
            'connected',
            'unacquired',
            'disconnected',
            'unavailable',
        ];

        entry.forEach((status) => {
            expect(dUtils.getStatusColor(status)).toMatchSnapshot();
        });
    });

    it('get status name', () => {
        const entry = [
            0,
            null,
            'sdsdsdsd',
            'used-in-other-window',
            'connected',
            'unacquired',
            'disconnected',
            'unavailable',
        ];

        entry.forEach((status) => {
            expect(dUtils.getStatusName(status)).toMatchSnapshot();
        });
    });
});
