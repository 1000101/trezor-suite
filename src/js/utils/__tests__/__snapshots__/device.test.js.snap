// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`device utils get status 1`] = `"Status unknown"`;

exports[`device utils get status 2`] = `"Status unknown"`;

exports[`device utils get status 3`] = `"Status unknown"`;

exports[`device utils get status 4`] = `"Status unknown"`;

exports[`device utils get status color 1`] = `"#494949"`;

exports[`device utils get status color 2`] = `"#494949"`;

exports[`device utils get status color 3`] = `"#494949"`;

exports[`device utils get status color 4`] = `"#EB8A00"`;

exports[`device utils get status color 5`] = `"#01B757"`;

exports[`device utils get status color 6`] = `"#EB8A00"`;

exports[`device utils get status color 7`] = `"#ED1212"`;

exports[`device utils get status color 8`] = `"#ED1212"`;

exports[`device utils get status name 1`] = `"Status unknown"`;

exports[`device utils get status name 2`] = `"Status unknown"`;

exports[`device utils get status name 3`] = `"Status unknown"`;

exports[`device utils get status name 4`] = `"Used in other window"`;

exports[`device utils get status name 5`] = `"Connected"`;

exports[`device utils get status name 6`] = `"Used in other window"`;

exports[`device utils get status name 7`] = `"Disconnected"`;

exports[`device utils get status name 8`] = `"Unavailable"`;
